File: cses/07_mathematics/1713_counting-divisors/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    nums = [int(v) for v in data[1:1 + n]]
    if not nums:
        return

    maxv = max(nums)

    # smallest prime factor sieve: spf[i] là ước nguyên tố nhỏ nhất của i.
    # Khởi tạo spf[i] = i, rồi với mỗi nguyên tố i đánh dấu các bội chưa chạm.
    spf = list(range(maxv + 1))
    i = 2
    while i * i <= maxv:
        if spf[i] == i:  # i là số nguyên tố
            step = i
            for j in range(i * i, maxv + 1, step):
                if spf[j] == j:
                    spf[j] = i
        i += 1

    # Với mỗi x: liên tục chia cho spf[x] để lấy số mũ e của từng nguyên tố,
    # số ước là tích các (e + 1).
    out = []
    append = out.append
    for x in nums:
        d = 1
        while x > 1:
            p = spf[x]
            e = 0
            while x % p == 0:
                x //= p
                e += 1
            d *= e + 1
        append(d)

    sys.stdout.write('\n'.join(map(str, out)) + '\n')


main()

File: cses/07_mathematics/1713_counting-divisors/solution.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> values(static_cast<size_t>(n));
    for (int& value : values) {
        cin >> value;
    }

    // Xây smallest prime factor sieve tới giá trị lớn nhất: spf[i] là ước
    // nguyên tố nhỏ nhất của i. Khởi tạo spf[i] = i, rồi với mỗi nguyên tố
    // prime, đánh dấu các bội chưa được chạm.
    const int maximum = *max_element(values.begin(), values.end());
    vector<int> smallest_prime(static_cast<size_t>(maximum) + 1);
    iota(smallest_prime.begin(), smallest_prime.end(), 0);
    for (int prime = 2; prime * prime <= maximum; ++prime) {
        // spf[prime] != prime nghĩa là prime đã có ước nhỏ hơn, không phải
        // nguyên tố, bỏ qua.
        if (smallest_prime[static_cast<size_t>(prime)] != prime) {
            continue;
        }
        for (int multiple = prime * prime; multiple <= maximum; multiple += prime) {
            // Chỉ gán khi bội chưa được nguyên tố nhỏ hơn đánh dấu.
            if (smallest_prime[static_cast<size_t>(multiple)] == multiple) {
                smallest_prime[static_cast<size_t>(multiple)] = prime;
            }
        }
    }

    // Với mỗi x: phân tích thừa số bằng cách liên tục chia cho spf[x], đếm số
    // mũ e của từng nguyên tố, rồi nhân số ước với (e + 1).
    for (int value : values) {
        int divisors = 1;
        while (value > 1) {
            const int prime = smallest_prime[static_cast<size_t>(value)];
            int exponent = 0;
            do {
                value /= prime;
                ++exponent;
            } while (value > 1 && smallest_prime[static_cast<size_t>(value)] == prime);
            divisors *= exponent + 1;
        }
        cout << divisors << '\n';
    }
}
